File: oe129bc-python/src/oe129bc/models/transaction.py
"""Pydantic model for the TRAN-RECORD layout from CVTRA05Y.cpy (350 bytes).

Field layout (COBOL copybook CVTRA05Y.cpy):
    01  TRAN-RECORD.
        05  TRAN-ID             PIC X(16).       -- Primary key, sequential
        05  TRAN-TYPE-CD        PIC X(02).       -- Transaction type code
        05  TRAN-CAT-CD         PIC 9(04).       -- Category code (numeric)
        05  TRAN-SOURCE         PIC X(10).       -- Source identifier
        05  TRAN-DESC           PIC X(100).      -- Description
        05  TRAN-AMT            PIC S9(09)V99.   -- Signed decimal amount
        05  TRAN-MERCHANT-ID    PIC 9(09).       -- Merchant identifier
        05  TRAN-MERCHANT-NAME  PIC X(50).       -- Merchant name
        05  TRAN-MERCHANT-CITY  PIC X(50).       -- Merchant city
        05  TRAN-MERCHANT-ZIP   PIC X(10).       -- Merchant zip
        05  TRAN-CARD-NUM       PIC X(16).       -- Card number
        05  TRAN-ORIG-TS        PIC X(26).       -- Origination timestamp
        05  TRAN-PROC-TS        PIC X(26).       -- Processing timestamp
        05  FILLER              PIC X(20).       -- Reserved
    Total: 350 bytes
"""

from decimal import Decimal
from typing import ClassVar

from pydantic import BaseModel, Field


class FieldSpec:
    """Describes a fixed-length COBOL field: name, offset, length, and type."""

    def __init__(self, name: str, offset: int, length: int, field_type: str = "text") -> None:
        self.name = name
        self.offset = offset
        self.length = length
        self.field_type = field_type  # "text", "numeric", "signed_decimal"


# Field layout matching CVTRA05Y.cpy byte offsets exactly
TRAN_RECORD_FIELDS: list[FieldSpec] = [
    FieldSpec("tran_id", 0, 16, "text"),
    FieldSpec("tran_type_cd", 16, 2, "text"),
    FieldSpec("tran_cat_cd", 18, 4, "numeric"),
    FieldSpec("tran_source", 22, 10, "text"),
    FieldSpec("tran_desc", 32, 100, "text"),
    FieldSpec("tran_amt", 132, 11, "signed_decimal"),
    FieldSpec("tran_merchant_id", 143, 9, "numeric"),
    FieldSpec("tran_merchant_name", 152, 50, "text"),
    FieldSpec("tran_merchant_city", 202, 50, "text"),
    FieldSpec("tran_merchant_zip", 252, 10, "text"),
    FieldSpec("tran_card_num", 262, 16, "text"),
    FieldSpec("tran_orig_ts", 278, 26, "text"),
    FieldSpec("tran_proc_ts", 304, 26, "text"),
    FieldSpec("filler", 330, 20, "text"),
]

RECORD_LENGTH = 350


class TransactionRecord(BaseModel):
    """Pydantic model for the TRAN-RECORD (CVTRA05Y.cpy, 350 bytes).

    All text fields are stored stripped of trailing whitespace.
    The amount field uses Decimal for exact arithmetic (no floating-point).
    """

    tran_id: str = Field(max_length=16, description="Primary key, sequential (PIC X(16))")
    tran_type_cd: str = Field(max_length=2, description="Transaction type code (PIC X(02))")
    tran_cat_cd: str = Field(max_length=4, description="Category code, numeric (PIC 9(04))")
    tran_source: str = Field(max_length=10, description="Source identifier (PIC X(10))")
    tran_desc: str = Field(max_length=100, description="Description (PIC X(100))")
    tran_amt: Decimal = Field(description="Signed decimal amount (PIC S9(09)V99)")
    tran_merchant_id: str = Field(max_length=9, description="Merchant identifier, numeric (PIC 9(09))")
    tran_merchant_name: str = Field(max_length=50, description="Merchant name (PIC X(50))")
    tran_merchant_city: str = Field(max_length=50, description="Merchant city (PIC X(50))")
    tran_merchant_zip: str = Field(max_length=10, description="Merchant zip (PIC X(10))")
    tran_card_num: str = Field(max_length=16, description="Card number (PIC X(16))")
    tran_orig_ts: str = Field(max_length=26, description="Origination timestamp (PIC X(26))")
    tran_proc_ts: str = Field(max_length=26, description="Processing timestamp (PIC X(26))")
    filler: str = Field(default="", max_length=20, description="Reserved (PIC X(20))")

    RECORD_LENGTH: ClassVar[int] = RECORD_LENGTH
    FIELDS: ClassVar[list[FieldSpec]] = TRAN_RECORD_FIELDS

    model_config = {"frozen": False, "str_strip_whitespace": True}

File: java-migration/src/main/java/com/carddemo/transaction/Transaction.java
package com.carddemo.transaction;

import java.math.BigDecimal;
import java.time.LocalDateTime;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * JPA entity mapping the COBOL {@code TRAN-RECORD} (copybook
 * {@code app/cpy/CVTRA05Y.cpy}, VSAM TRANSACT KSDS, record length 350).
 *
 * <p>Field mappings preserve the original COBOL field names and PIC clauses for
 * traceability. The 20-byte trailing {@code FILLER} is intentionally not mapped.
 * {@code TRAN-AMT} uses {@link BigDecimal} (never floating point) to preserve the
 * COBOL fixed-point {@code S9(09)V99} semantics — note this is an 11-byte zoned
 * field (9 integer + 2 fraction), narrower than the 12-byte Account money fields.
 * The two timestamp fields use {@link LocalDateTime}; their stored form is
 * {@code yyyy-MM-dd HH:mm:ss.SSSSSS} (see {@link TransactionSeedLoader#TS_FORMAT}).</p>
 */
@Entity
@Table(name = "transaction")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Transaction {

    /** COBOL: {@code TRAN-ID PIC X(16)} — VSAM primary key (offset 0, len 16). */
    @Id
    @Column(name = "tran_id", length = 16)
    private String tranId;

    /** COBOL: {@code TRAN-TYPE-CD PIC X(02)} — transaction type code (offset 16, len 2). */
    @Column(name = "tran_type_cd", length = 2)
    private String tranTypeCd;

    /** COBOL: {@code TRAN-CAT-CD PIC 9(04)} — transaction category code (offset 18, len 4). */
    @Column(name = "tran_cat_cd")
    private Integer tranCatCd;

    /** COBOL: {@code TRAN-SOURCE PIC X(10)} — origin channel (offset 22, len 10). */
    @Column(name = "tran_source", length = 10)
    private String tranSource;

    /** COBOL: {@code TRAN-DESC PIC X(100)} — free-text description (offset 32, len 100). */
    @Column(name = "tran_desc", length = 100)
    private String tranDesc;

    /** COBOL: {@code TRAN-AMT PIC S9(09)V99} — signed amount, 11-byte zoned decimal (offset 132, len 11). */
    @Column(name = "tran_amt", precision = 11, scale = 2)
    private BigDecimal tranAmt;

    /** COBOL: {@code TRAN-MERCHANT-ID PIC 9(09)} — merchant id (offset 143, len 9). */
    @Column(name = "tran_merchant_id")
    private Long tranMerchantId;

    /** COBOL: {@code TRAN-MERCHANT-NAME PIC X(50)} — merchant name (offset 152, len 50). */
    @Column(name = "tran_merchant_name", length = 50)
    private String tranMerchantName;

    /** COBOL: {@code TRAN-MERCHANT-CITY PIC X(50)} — merchant city (offset 202, len 50). */
    @Column(name = "tran_merchant_city", length = 50)
    private String tranMerchantCity;

    /** COBOL: {@code TRAN-MERCHANT-ZIP PIC X(10)} — merchant ZIP (offset 252, len 10). */
    @Column(name = "tran_merchant_zip", length = 10)
    private String tranMerchantZip;

    /** COBOL: {@code TRAN-CARD-NUM PIC X(16)} — card number, XREF lookup key (offset 262, len 16). */
    @Column(name = "tran_card_num", length = 16)
    private String tranCardNum;

    /** COBOL: {@code TRAN-ORIG-TS PIC X(26)} — origination timestamp (offset 278, len 26). */
    @Column(name = "tran_orig_ts")
    private LocalDateTime tranOrigTs;

    /** COBOL: {@code TRAN-PROC-TS PIC X(26)} — processing/posting timestamp (offset 304, len 26). */
    @Column(name = "tran_proc_ts")
    private LocalDateTime tranProcTs;
}
